File: Lab 6 - RSA ElGamal/RSAServer.cpp
#include <cstring>
#include <iostream>
#include <netinet/in.h>
#include <sys/socket.h>
#include <bits/stdc++.h>
#include <unistd.h>
using namespace std;

#define PUB_P 7919
#define PUB_Q 1009

int clientSocket, serverSocket;

void serverConnect() {
    serverSocket = socket(AF_INET, SOCK_STREAM, 0);

    sockaddr_in serverAddress;
    serverAddress.sin_family = AF_INET;
    serverAddress.sin_port = htons(8080);
    serverAddress.sin_addr.s_addr = INADDR_ANY;

    bind(serverSocket, (struct sockaddr *)&serverAddress, sizeof(serverAddress));

    listen(serverSocket, 5);

    clientSocket = accept(serverSocket, nullptr, nullptr);
}

void sendData(string message) {
    char buffer[100] = {0};
    copy(message.begin(), message.end(), buffer);
    send(clientSocket, buffer, sizeof(buffer), 0);
}

string receiveData() {
    char buffer[100] = {0};
    recv(clientSocket, buffer, sizeof(buffer), 0);
    string message = string(buffer);
    return message;
}

int extended_euclidean(int num, int phi, int& x, int& y) {
    if (phi == 0) {
        x = 1, y = 0;
        return num;
    }
    int x1, y1;
    int d = extended_euclidean(phi, num % phi, x1, y1);
    y = x1 - (num / phi) * y1;
    x = y1;
    return d;
}

int gcd(int a, int b)
{
    if(b==0) return a;
    return gcd(b, a%b);
}

int find_coprime(int phi)
{
    for(int i=2; i<phi; i++){
        int gcd_phi = gcd(i, phi);
        if (gcd_phi == 1) return i;
    }
    return 1;
}

int generate_key(int& n)
{
    n = PUB_P * PUB_Q;
    int phi = (PUB_P-1)*(PUB_Q-1);
    int e = find_coprime(phi);
    int inv, y;
    int d = extended_euclidean(e, phi, inv, y);
    if(inv<0) inv = (inv%phi + phi) % phi;
    return inv;
}

int decrypt (int ct, int key, int n)
{
    int res = 1;
    ct = ct % n;
    while (key>0){
      if(key & 1) res = (res * 1LL * ct) % n;
      ct = (ct * 1LL * ct) % n;
      key = key/2;
    }
    return res;
}

int main() {
    serverConnect();

    sendData("Server Online!");
    cout << ">: " << receiveData() << endl;
    
    int n;
    int d = generate_key(n);
    cout << ">: Private key(d, n): (" << d << ", " << n << ")" << endl;
    
    string input_ciphertext = receiveData();
    int ciphertext = stoi(input_ciphertext);
    cout << ">: received text: " << ciphertext<<endl;
    
    int plaintext = decrypt(ciphertext, d, n);
    cout << ">: Decrypted Text: " << plaintext << endl;
    
    cout << ">: Server Exiting!!" << endl << endl;
    close(serverSocket);
    
    return 0;
}

File: Lab 6 - RSA ElGamal/RSAServer.py
import socket

def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

def mod_inverse(e, phi):
    d = 0
    x1, x2, y1 = 0, 1, 1
    temp_phi = phi

    while e > 0:
        temp1 = temp_phi // e
        temp2 = temp_phi - temp1 * e
        temp_phi, e = e, temp2
        x = x2 - temp1 * x1
        y = d - temp1 * y1
        x2, x1 = x1, x
        d, y1 = y1, y

    return d + phi if d < 0 else d

def generate_keys():
    p = 61
    q = 53
    n = p * q
    phi = (p - 1) * (q - 1)
    e = 17
    while gcd(e, phi) != 1:
        e += 2
    d = mod_inverse(e, phi)
    return ((e, n), (d, n))

def decrypt(cipher, priv_key):
    d, n = priv_key
    decrypted = [chr(pow(c, d, n)) for c in cipher]
    return ''.join(decrypted)

host = '127.0.0.1'
port = 12345
server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server.bind((host, port))
server.listen(1)
print(f"Server started. Waiting for connection at {host}:{port}...")

conn, addr = server.accept()
print(f"Connected by {addr}")

public, private = generate_keys()
print(f"Public key: {public}")
print(f"Private key: {private}")

# Send the public key to the client
conn.send(f"{public[0]},{public[1]}".encode())

data = conn.recv(4096).decode()
print(data)
cipher = list(map(int, data.split(',')))

plain_text = decrypt(cipher, private)
print("Decrypted message:", plain_text)

conn.close()
server.close()
